src/tasks/gitea/gitea-drone-oauth.ts

```diff
@@ -0,0 +1,212 @@
+import * as k8s from '@kubernetes/client-node'
+import axios, { AxiosRequestConfig } from 'axios'
+import querystring from 'querystring'
+
+import { UserApi, CreateOAuth2ApplicationOptions, OAuth2Application } from '@redkubes/gitea-client-node'
+
+import { cleanEnv, GITEA_USER, GITEA_PASSWORD, GITEA_URL, GITEA_REPO } from '../../validators'
+
+const env = cleanEnv({
+  GITEA_USER,
+  GITEA_PASSWORD,
+  GITEA_URL,
+  GITEA_REPO,
+})
+type DroneSecret = {
+  clientId: string
+  clientSecret: string
+}
+class GiteaDroneOAuth {
+  apiClient: k8s.CoreV1Api
+  k8s_namespace = 'gitea'
+  k8s_secret_name = 'gitea-drone-secret'
+  csrfCookieName = '_csrf'
+
+  oauthData: DroneSecret
+
+  giteaUrl: string
+  droneUrl: string
+
+  csrfToken: string
+  constructor() {
+    this.giteaUrl = env.GITEA_URL
+    if (this.giteaUrl.endsWith('/')) {
+      this.giteaUrl = this.giteaUrl.slice(0, -1)
+    }
+    this.droneUrl = this.giteaUrl.replace('gitea', 'drone')
+
+    this.main()
+  }
+
+  getApiClient(): k8s.CoreV1Api {
+    if (this.apiClient) return this.apiClient
+    const kc = new k8s.KubeConfig()
+    kc.loadFromDefault()
+    this.apiClient = kc.makeApiClient(k8s.CoreV1Api)
+    return this.apiClient
+  }
+
+  async secretExists(): Promise<DroneSecret> {
+    const client = this.getApiClient()
+    try {
+      const response = (await client.readNamespacedSecret(this.k8s_secret_name, this.k8s_namespace)).body
+      return response.data as DroneSecret
+    } catch (e) {
+      return undefined
+    }
+  }
+  async authorizeOAuth() {
```

Comment:
The function name does not really reflect what it performs.
It is also very complex thus not testable.
Please remember about single responsibility principle and think if some parts can be extracted to smaller functions.

Reply:
But it's close: `authorizeOAuthApp` would be better